theseus/extlib/cusolver_lu_solver.cpp

```diff
@@ -0,0 +1,334 @@
+                                                   luBuffer.data_ptr<uint8_t>()));
+
+        // P, Q (for A's factorization) are obtained as composition of permutations
+        P_cpu = torch::empty(rows, torch::TensorOptions(torch::kInt));
+        Q_cpu = torch::empty(cols, torch::TensorOptions(torch::kInt));
+        int* pP = P_cpu.data_ptr<int>(), *pP_lu = P_lu.data_ptr<int>();
+        int* pQ = Q_cpu.data_ptr<int>(), *pQ_lu = Q_lu.data_ptr<int>();
+        for(int j = 0; j < rows; j++){
+            pP[j] = pQperm[pP_lu[j]];
+        }
+        for(int j = 0; j < cols; j++){
+            pQ[j] = pQperm[pQ_lu[j]];
+        }
+
+        CUSOLVER_CHECK(cusolverSpDestroyCsrluInfoHost(info));
+        TORCH_CUDASPARSE_CHECK(cusparseDestroyMatDescr(A_descr));
+    }
+
+    // cusolverRf part
+    const cusolverRfFactorization_t fact_alg = CUSOLVERRF_FACTORIZATION_ALG0; // default
+    const cusolverRfTriangularSolve_t solve_alg = CUSOLVERRF_TRIANGULAR_SOLVE_ALG1; // default  
+    double nzero = 0.0;
+    double nboost = 0.0;
+    CUSOLVER_CHECK(cusolverRfCreate(&cusolverRfH));
+    CUSOLVER_CHECK(cusolverRfSetNumericProperties(cusolverRfH, nzero, nboost));
+    CUSOLVER_CHECK(cusolverRfSetAlgs(cusolverRfH, fact_alg, solve_alg));
+    CUSOLVER_CHECK(cusolverRfSetMatrixFormat(cusolverRfH, CUSOLVERRF_MATRIX_FORMAT_CSR, CUSOLVERRF_UNIT_DIAGONAL_ASSUMED_L));
+    CUSOLVER_CHECK(cusolverRfSetResetValuesFastMode(cusolverRfH, CUSOLVERRF_RESET_VALUES_FAST_MODE_ON));
+
+    at::Tensor A_val_cpu = torch::empty(batchSize * nnz, torch::TensorOptions(torch::kDouble));
+    at::Tensor A_val_array_cpu = torch::empty(batchSize * sizeof(double*), torch::TensorOptions(torch::kByte));
+    double* pA_val_cpu = A_val_cpu.data_ptr<double>();
+    double** pA_val_array_cpu = (double**)A_val_array_cpu.data_ptr<uint8_t>();
+    for(int i = 0; i < batchSize; i++) {
+        pA_val_array_cpu[i] = pA_val_cpu + nnz * i;
+    }
+
+    CUSOLVER_CHECK(cusolverRfBatchSetupHost(batchSize,
+                                            rows, nnz,
+                                            A_rowPtr_cpu.data_ptr<int>(), A_colInd_cpu.data_ptr<int>(), pA_val_array_cpu,
+                                            L_nnz, L_rowPtr.data_ptr<int>(), L_colInd.data_ptr<int>(), L_val.data_ptr<double>(),
+                                            U_nnz, U_rowPtr.data_ptr<int>(), U_colInd.data_ptr<int>(), U_val.data_ptr<double>(),
+                                            P_cpu.data_ptr<int>(), Q_cpu.data_ptr<int>(),
+                                            cusolverRfH));
+
+    CUSOLVER_CHECK(cusolverRfBatchAnalyze(cusolverRfH));
+
+    P = P_cpu.cuda();
+    Q = Q_cpu.cuda();
+}
+
+std::vector<int> CusolverLUSolver::factor(const torch::Tensor& A_val) {
+
+    TORCH_CHECK(A_val.device().is_cuda());
+    TORCH_CHECK(A_val.dim() == 2);
+    TORCH_CHECK(A_val.size(0) <= batchSize); // currently failing unless "==", cublas bug?!
```

Comment:
Weird. Maybe a work around can be to check both <= and >= ?

Reply:
oh yeah will update the check - the comment is that it's sad that cublas forces the exact batch size.

Reply:
so what I would like to say is that we would like to check "<=", but cuda functions fail unless "==" holds. Probably we should check "==" anyway, untill an improved cuda is released.